### DISCUSSION TITLE
Watch face on WearOS 5

### DISCUSSION THREAD

**Discussion by zyt** [j-kaltes/Juggluco#187] [General]
Hi,

I was about to buy Galaxy Watch 7 which comes pre-installed with WatchOS 5. Then I noticed that in the download section of the Juggluco [website](https://www.juggluco.nl/Juggluco/download.html) there is a warning that watch face does not work with WearOS 5. 

I read that WearOS 5 discontinued support for old watch face formats and going forward they will no longer be supported. Does it mean that Juggluco installed on WearOS 5 watch will not be able to display the glucose levels on the watch? Is there perhaps some workaround or will it require code rewrite to support Watch Face Format (WFF)?

**Comment by j-kaltes** [comment]
Programmed watch faces are verboten under WearOS 5. Juggluco’s watch face can’t be implementing as purely formatting, so it can’t be implemented within the WearOS 5 restrictions. There is an adb command to turn off some of the WearOS 5 restrictions for testing purposes. Maybe the watch face can be used after that command.
```bash
adb shell device_config put wear_frameworks \
  com.google.android.clockwork.systemui.flags.ambiactive_components_expirable \
  false && adb reboot
```
See: https://developer.android.com/training/wearables/versions/5/changes#always-on-background-test
An alternative is to use _Floating Glucose_ above another watch face or to make use of a glucose complication inside a complication slot of another watch face, see  https://github.com/pachi81/GlucoDataHandler/releases

**Reply by zyt** [reply]
Thank you for fast reply! 

I apologize if my questions are a little bit unclear. I'm still new to the WearOS ecosystem. 

Hmm, this makes direct connection between WearOS 5 smartwatch and CGM sensor a little bit tricky but nonetheless possible?
Apparently, GlucoDataHandler can run in standalone mode on the watch as confirmed by main dev [here](https://insulinclub.de/index.php?thread/34766-zeigt-her-eure-watchfaces/&postID=751199#post751199) so it should be possible to use the glucose complication as you suggested.

I'm going to proceed with buying GW7 and I will report back in coming weeks.

**Comment by zyt** [comment]
I successfully sideloaded Juggluco on Galaxy Watch7 (GW7) via ADB and I can confirm that phone-watch connection is super stable.

As suggested by @j-kaltes I installed [GlucoDataHandler](https://github.com/pachi81/GlucoDataHandler/) on GW7 and used the complications from GlucoDataHandler to display glucose values on a default watchface. I can confirm this is working very well even on WearOS 5 and with direct watch-sensor connection.

Direct watch-sensor connection does work, however there are constant disconnects between the watch and the sensor (either status 133 or status 19) and as a result the glucose values are transmitted only every 2 minutes instead of every 1 minute. I read in other thread regarding GW6 where people observed similar behaviour. Is there any way to somehow overcome this problem?
I tried enabling Turbo in sensor settings and also `adb shell dumpsys deviceidle whitelist +tk.glucodata` but both with no effect.

**Reply by j-kaltes** [reply]
Did you try:
```bash
adb shell cmd appops set tk.glucodata RUN_IN_BACKGROUND allow
```
?
See [Left menu->Sensor->Doze mode?](https://www.juggluco.nl/Jugglucohelp/battery.html)
Did you try if the watch face can be used after 
```bash
adb shell device_config put wear_frameworks \
  com.google.android.clockwork.systemui.flags.ambiactive_components_expirable \
  false && adb reboot
```
?

**Reply by zyt** [reply]
Yes background activity is enabled (via galaxy watch companion app) but I did run the adb command for good measure as well but it has no effect either. Glucose values are received only every 2 minutes even if I have Juggluco app open on the GW7 and looking at the Sensor information without screen turning off.

I did try running the adb command to enable the watch face but sadly it has no effect.

**Reply by j-kaltes** [reply]
You use Libre 3 sensors?

**Reply by zyt** [reply]
Sorry forgot to mention it, currently I'm using Freestyle Libre 2 bought in Luxembourg so I suppose that is the EU version.

**Reply by j-kaltes** [reply]
Can you run the logging version of Juggluco on your watch and send trace.log to me? See: https://www.juggluco.nl/Juggluco/download.html#bugreport

I didn't hear of a once in two minute connection on WearOS galaxy watch with Libre 2.

**Reply by zyt** [reply]
Sure I can do that, how do switch from normal to debug version and retain the app data? Is it `adb install -r <pkg_name>`?

**Reply by j-kaltes** [reply]
yes, just install the apk with adb as usual. The data will be retained as with any other update.

**Reply by zyt** [reply]
@j-kaltes I just sent you the log file via email.

**Comment by j-kaltes** [comment]
Your log shows 5 seconds after every received glucose value a status=19 connection error.
Can you send me also a trace.log from your phone? To see that is is not something of the sensor.

The slowness of the watch, could make it different, but on the (also slow) Watch 4 I don’t have these connection errors. Maybe it is caused be a kind of misplaced attempt to save battery power, which in reality causes more battery usage because of the reconnection attempts.

**Reply by zyt** [reply]
That is very interesting since I disabled all the known battery optimisation. I sent you the trace logs from both phone and watch.

**Reply by j-kaltes** [reply]
The trace.log from the phone shows exactly one glucose value received directly from the sensor. I wanted to know how it receives more than one value.
Disabling battery optimizations also don't allow you to make use of programmed watch faces.

**Reply by zyt** [reply]
Oh I'm sorry I misunderstood, I will resend a log where the sensor is connected to my phone.

**Reply by zyt** [reply]
Re-send the log file to you.

**Reply by j-kaltes** [reply]
I see, yes there is works without these status=19 connection errors. So the code is right and the watch wrong.

**Reply by zyt** [reply]
Fair enough, unfortunately I highly doubt that Samsung would be willing to do anything about this. Is there anything else I could do from my end to debug the issue?

**Reply by Puderzo** [reply]
Just a quick note on this, my son's Galaxy Watch 6 exhibits the same symptoms of 2-3 minutes between values. Doesn't matter if it's Libre 2 or 3. I seem to remember that the same was true also after we updated his previous Galaxy Watch 4 Classic to Wear OS 4, so I think this is some kind of attempt to limit bluetooth usage in the name of battery optimization.

**Reply by j-kaltes** [reply]
I have Watch 4 updated to WearOS 4.0 and Juggluco on that watch receives a glucose value every minute from Libre 2 sensors.
![everyminute](https://github.com/user-attachments/assets/8c6d0714-2501-4537-855c-ef5fcaaef71f)
![WearOS4 0](https://github.com/user-attachments/assets/0e916651-ed4c-467e-94a9-69e68d29cdac)

**Reply by Puderzo** [reply]
My bad, might have been just the change of watch then. Perhaps we should go back to using the Watch 4 Classic instead then :) The reason we got the Watch 6 was we wanted better battery life, but that has yet to materialize, instead it appears to sometimes be worse, even though it has a larger battery and should be "better optimized".

**Reply by zyt** [reply]
I managed to get my hands on Libre 3 (DE version). I tested direct watch-sensor connection and I am able to get glucose values every minute as long as the connection remains stable (no interference in 2.4Ghz range)...
